lib/std/interface.zig

```diff
@@ -0,0 +1,478 @@
+    }
+}
+
+fn selfPtrAs(self: *SelfType, comptime T: type) *T {
+    if (@sizeOf(T) > 0) {
+        return @alignCast(@alignOf(T), @ptrCast(*align(1) T, self));
+    } else {
+        return undefined;
+    }
+}
+
+fn constSelfPtrAs(self: *const SelfType, comptime T: type) *const T {
+    if (@sizeOf(T) > 0) {
+        return @alignCast(@alignOf(T), @ptrCast(*align(1) const T, self));
+    } else {
+        return undefined;
+    }
+}
+
+pub const Storage = struct {
+    pub const NonOwning = struct {
+        erased_ptr: *SelfType,
+
+        pub fn init(args: var) !NonOwning {
+            if (args.len != 1) {
+                @compileError("NonOwning storage expected a 1-tuple in initialization.");
+            }
+
+            return NonOwning{
+                .erased_ptr = makeSelfPtr(args.@"0"),
+            };
+        }
+
+        pub fn getSelfPtr(self: NonOwning) *SelfType {
+            return self.erased_ptr;
+        }
+
+        pub fn deinit(self: NonOwning) void {}
+    };
+
+    pub const Owning = struct {
+        allocator: *Allocator,
+        erased_ptr: *SelfType,
+
+        pub fn init(args: var) !Owning {
+            if (args.len != 2) {
+                @compileError("Owning storage expected a 2-tuple in initialization.");
+            }
+
+            const AllocT = @TypeOf(args.@"0");
+
+            var mem = try args.@"1".create(AllocT);
+            mem.* = args.@"0";
+
+            return Owning{
+                .allocator = args.@"1",
+                .erased_ptr = makeSelfPtr(mem),
+            };
+        }
+
+        pub fn getSelfPtr(self: Owning) *SelfType {
+            return self.erased_ptr;
+        }
+
+        pub fn deinit(self: Owning) void {
+            self.allocator.destroy(selfPtrAs(self.erased_ptr, u8));
+        }
+    };
+
+    pub fn Inline(comptime size: usize) type {
+        return struct {
+            const Self = @This();
+
+            mem: [size]u8,
+
+            pub fn init(args: var) !Self {
+                if (args.len != 1) {
+                    @compileError("Inline storage expected a 1-tuple in initialization.");
+                }
+
+                const ImplSize = @sizeOf(@TypeOf(args.@"0"));
+
+                if (ImplSize > size) {
+                    @compileError("Type does not fit in inline storage.");
+                }
+
+                var self: Self = undefined;
+
+                std.mem.copy(u8, self.mem[0..], @ptrCast([*]const u8, &args.@"0")[0..ImplSize]);
+                return self;
+            }
+
+            pub fn getSelfPtr(self: Self) *SelfType {
+                return makeSelfPtr(&self.mem[0]);
+            }
+
+            pub fn deinit(self: Self) void {}
+        };
+    }
+
+    pub fn InlineOrOwning(comptime size: usize) type {
+        return struct {
+            const Self = @This();
+
+            // TODO: Pack this tightly with an extern union?
+            // Check resulting size of this union.
+            data: union {
```

Comment:
This should be a tagged union (`union(enum)`) instead of a untagged one, as

> Unions can be declared with an enum tag type. This turns the union into a tagged union, which makes it eligible to use with switch expressions.

You are using it below in a switch statement